core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/pythonUDF/PythonUDFOpDesc.java

```diff
@@ -69,21 +91,32 @@ public OperatorInfo operatorInfo() {
     @Override
     public Schema getOutputSchema(Schema[] schemas) {
         Schema inputSchema = schemas[0];
-        if (inputColumns != null && !inputColumns.isEmpty()) {
+
+        // check if inputColumns are presented in inputSchema.
+        if (inputColumns != null) {
             for (String s : inputColumns) {
-                if (!inputSchema.containsAttribute(s)) throw new RuntimeException("No such column:" + s);
+                if (!inputSchema.containsAttribute(s)) throw new RuntimeException("No such column:" + s + ".");
             }
         }
 
-        Schema outputSchema = inputSchema;
-        if (outputColumns != null && !outputColumns.isEmpty()) {
+        Schema.Builder outputSchemaBuilder = Schema.newBuilder();
+        if (pythonUDFType != Training) {
```

Comment:
I see you also added map and filter type. "map" type might correspond to the case where a new column is added. but filter type is not handled here. We don't have a filter udf yet, but it's better to make the code correct now, since we might forget later.

Reply:
both `map` and `filter` are keeping the same schema from the input. they both can append new attributes. map usually does, and filter usually does not append new attributes, but still possible. I will add some comments here.